pydap_extras/handlers/sql.py

```diff
@@ -0,0 +1,409 @@
+        ...     'database': { 'dsn': 'sqlite:///test.db', 'table': 'test', 'order': 'idx' },
+        ...     'index': { 'col': 'idx' },
+        ...     'temperature': { 'col': 'temperature' },
+        ...     'site': { 'col': 'site' }}
+        >>> seq = SequenceType('example')
+        >>> seq['index'] = BaseType('index')
+        >>> seq['temperature'] = BaseType('temperature')
+        >>> seq['site'] = BaseType('site')
+        >>> seq.data = SQLData(config, seq.id, ('index', 'temperature', 'site'),
+        ...     {'index': np.int32, 'temperature': np.float32, 'site': np.dtype('|S14')})
+        >>> for line in seq:
+        ...     print line
+        (10.0, 15.2, u'Diamond_St')
+        (11.0, 13.1, u'Blacktail_Loop')
+        (12.0, 13.3, u'Platinum_St')
+        (13.0, 12.1, u'Kodiak_Trail')
+        >>> for line in seq['temperature', 'site', 'index']:
+        ...     print line
+        (15.2, u'Diamond_St', 10.0)
+        (13.1, u'Blacktail_Loop', 11.0)
+        (13.3, u'Platinum_St', 12.0)
+        (12.1, u'Kodiak_Trail', 13.0)
+    We can iterate over children:
+        >>> for line in seq['temperature']:
+        ...     print line
+        15.2
+        13.1
+        13.3
+        12.1
+    We can filter the data:
+        >>> for line in seq[ seq.index > 10 ]:
+        ...     print line
+        (11.0, 13.1, u'Blacktail_Loop')
+        (12.0, 13.3, u'Platinum_St')
+        (13.0, 12.1, u'Kodiak_Trail')
+        >>> for line in seq[ seq.index > 10 ]['site']:
+        ...     print line
+        Blacktail_Loop
+        Platinum_St
+        Kodiak_Trail
+        >>> for line in seq['site', 'temperature'][ seq.index > 10 ]:
+        ...     print line
+        (u'Blacktail_Loop', 13.1)
+        (u'Platinum_St', 13.3)
+        (u'Kodiak_Trail', 12.1)
+    Or slice it:
+        >>> for line in seq[::2]:
+        ...     print line
+        (10.0, 15.2, u'Diamond_St')
+        (12.0, 13.3, u'Platinum_St')
+        >>> for line in seq[ seq.index > 10 ][::2]['site']:
+        ...     print line
+        Blacktail_Loop
+        Kodiak_Trail
+        >>> for line in seq[ seq.index > 10 ]['site'][::2]:
+        ...     print line
+        Blacktail_Loop
+        Kodiak_Trail
+    """
+
+    def __init__(
+        self,
+        config,
+        id,
+        cols,
+        dtypes,
+        template,
+        imap=None,
+        selection=None,
+        slice_=None,
+        level=0,
+    ):
+        self.template = template
+        self.config = config
+        self.id = id
+        self.cols = cols
+        self.dtypes = dtypes
+        self.selection = selection or []
+        self.slice = slice_ or (slice(None),)
+        self.level = level
+        self.imap = imap or []
+
+        # mapping between variable names and their columns
+        self.mapping = {
+            key: config[key]["col"] for key in config if "col" in config[key]
+        }
+
+    @property
+    def dtype(self):
+        peek = next(iter(self))
```

Comment:
Aren't the types *already* available in the `self.dtypes` property above (line 261)? Additionally I'd be kind of nervous about this method of type peeking since it executes the entire data query (line 317), which could be *substantial* and then doesn't use the results. Basically you'll get a 100% performance penalty on large queries.